ai_edge_torch/generative/layers/experimental/ekv_cache.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2024 The AI Edge Torch Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+# Cache management utilities.
+
+from dataclasses import dataclass
+from typing import List, Tuple
+
+import torch
+import torch.utils._pytree as pytree
+
+from ai_edge_torch import hlfb
+from ai_edge_torch.generative.layers import model_config
+
+
+@dataclass
+class KVCacheEntry:
+  """A single cache entry include K and V caches.
+
+  Returns:
+      _type_: _description_
+  """
+
+  k_cache: torch.Tensor
+  v_cache: torch.Tensor
+
+  @classmethod
+  def from_model_config(
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "KVCacheEntry":
+    shape = (
+        1,
+        config.kv_cache_max,
+        config.attn_config.num_query_groups,
+        config.head_dim,
+    )
+    k = torch.zeros(shape, dtype=dtype, device=device)
+    v = torch.zeros(shape, dtype=dtype, device=device)
+    obj = cls(k_cache=k, v_cache=v)
+    return obj
+
+
+@dataclass
+class EKVCache:
+  """A utility class for holding KV cach entries per layer."""
+
+  caches: Tuple[KVCacheEntry]
+
+  @classmethod
+  def from_model_config(
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "EKVCache":
+    """Build an instance of the class based on model config.
+    Args:
+        config (ModelConfig): Model config used for building the cache.
+        dtype (torch.dtype, optional): The data type of the cache tensor.
+          Defaults to torch.float32.
+        device (torch.device, optional): The device placement of the cache
+          tensors. Defaults to None.
+    Returns:
+        EKVCache: The created cache object.
+    """
+    caches = [
+        KVCacheEntry.from_model_config(config, dtype, device)
+        for _ in range(config.num_layers)
+    ]
+    obj = cls(caches=tuple(caches))
+    return obj
+
+
+def _flatten_kvc(kvc: EKVCache) -> Tuple[List, List]:
```

Comment:
can we add some unit tests for functions in this file?

Reply:
this will also help readers know what those functions are doing by looking at the tests

Reply:
Added some unit tests that check object serialization. However, please note that these are registered to pytorch export module so it has to go through that and these functions would not be tested directly by themselves.